---
title: "Peak to trough analysis"
author: "Scott G. Daniel, PennCHOP microbiome program"
date: \today
geometry: margin=3cm
output: 
    pdf_document:
        keep_tex: true
        toc: true
        toc_depth: 3
---

<!-- knitr setup -->
```{r knitr setup, echo=FALSE}
### ================
###   knitr setup
### ================
library(knitr)
knitr::opts_chunk$set(
	echo = FALSE,
	fig.width=8,
  fig.height=8,
	fig.align = "center",
	message = FALSE,
	warning = FALSE,
	cache = F,
	dpi = 100,
	tidy = FALSE
)
```
<!-- End, knitr setup -->

<!-- R packages -->
```{r R packages, message=FALSE, warning=FALSE}
### ================
###   R packages
### ================
#easier file path getting (NEVER use root_dir or setwd() AGAIN!)
library(here)
#This package will also help us more easily manipulate our data
library(tidyverse)
#So we can use %$% too
library(magrittr)
#for melt and cast in case we don't want to use gather and spread
library(reshape2)
#used often for diversity measurments
library(usedist)
library(vegan)
library(ape)
#graphing for greatness (and Faber!)
library(ggbeeswarm)
library(grid)
library(pheatmap)
library(RColorBrewer)
library(viridis)
#Süß tables
library(pander)
#We keep a lot of our metadata in Qiime-compatible format
library(qiimer)
library(lme4)
library(reshape2)
library(FactoMineR)

```
<!-- End, R packages -->

<!-- user defined functions -->
```{r functions}

source(here::here("common_functions.R"))

```
<!-- End, user defined functions -->

```{r some setup, include=FALSE}

data_dir <- here::here("peak_trough","Output")
plot_dir <- here::here("peak_trough","Output")

dir.create(plot_dir, recursive = T, showWarnings = F)

### mapping file path
mapping_file_fp <- here::here("Study_metadata.tsv")
s <- read.delim(mapping_file_fp)

```

# Introduction

Peak-to-through (PTR) is the analysis of DNA coverage of a genome to infer the growth rate of bacteria. The first paper on it by [Segal paper first author?] was limited in that it required fully assembled and annotated genomes. Additionally, the software used is based on a complex environment of out-of-date python libraries. Therefore, we use two other programs by [author, date] and [author, date] to estimate PTR values; both of these programs have the benefit of working on draft metagenomes.

# Peak-to-trough analysis from DEMIC

DEMIC (https://www.nature.com/articles/s41592-018-0182-0) follows this pipeline:

1. Calculates sequencing coverage in sliding windows over contigs

2. Filters contigs and samples that have low coverage

3. Uses linear mixed-models to correct GC bias

4. Uses PCA to infer relative distance of contigs to origin of replication

5. Performs linear regression of log-coverage to contigs and computes estimated peak-to-trough ratio

Here are the log-coverage to contigs graphs for each sample:

```{r processing, include=FALSE}

load(here("finished_demic.RData"))

#we re-run this to make sure we have all of the data frames

if( length(levels(X$contig)) >= 20 & length(levels(X$sample)) >= 3 ){
  cor_cutoff <- 0.98
  max_cor <- 0
  for (s2 in 1:3) {
    if(s2 == 2) {
      if(max_cor < 0.9) {
        break
      } else if (max_cor < 0.95) {
        cor_cutoff <- 0.95
      }
    }
    #print("s2:")
    #print(s2)
    nrm <- floor(length(levels(Y$contig))/5)
    set.seed(s2)
    designateR2 <- sample.int(10000,size=length(levels(Y$contig)),replace=FALSE)
    ContigDesignateR2 <- data.frame("Contig"=levels(Y$contig), "number"=designateR2)
    ContigDesignateRSort2 <- ContigDesignateR2[order(ContigDesignateR2[,2]),]
    nacontig_id <- NULL
    for (x in 1:4){
      for (y in (x+1):5) {
        if(x%in%nacontig_id) {
          break
        }
        if(y%in%nacontig_id) {
          next
        }
        ContigDesignateRemove1 <- ContigDesignateRSort2[(nrm*(x-1)+1):(nrm*x),1]
        ContigDesignateRemove2 <- ContigDesignateRSort2[(nrm*(y-1)+1):(nrm*y),1]

        pipelineX1 <- itePipelines(X[!X$contig%in%ContigDesignateRemove1,])
        if(length(pipelineX1) == 1) {
          break
        }
        pipelineX2 <- itePipelines(X[!X$contig%in%ContigDesignateRemove2,])
        if(length(pipelineX2) == 1) {
          nacontig_id <- c(nacontig_id, y)
          next
        }
        Samples_filteredXrm1 <- pipelineX1[[1]]
        summeryMeanYSortFilteredSampleContigXrm1 <- pipelineX1[[2]]
        contigPCAPC1FilteredXrm1 <- pipelineX1[[3]]
        rangeXrm1 <- pipelineX1[[4]]

        Samples_filteredXrm2 <- pipelineX2[[1]]
        summeryMeanYSortFilteredSampleContigXrm2 <- pipelineX2[[2]]
        contigPCAPC1FilteredXrm2 <- pipelineX2[[3]]
        rangeXrm2 <- pipelineX2[[4]]
        if( length(contigPCAPC1FilteredXrm1$contig) - length(intersect(contigPCAPC1FilteredXrm1$contig,contigPCAPC1FilteredXrm2$contig)) < 3 | length(contigPCAPC1FilteredXrm2$contig) - length(intersect(contigPCAPC1FilteredXrm1$contig,contigPCAPC1FilteredXrm2$contig)) < 3 ){
          next
        }

        SampleCorrectYWithPC1 <- merge( dcast(subset(summeryMeanYSortFilteredSampleContigXrm1,select=c("sample", "contig", "correctY")), contig ~ sample), contigPCAPC1FilteredXrm1 )

        lmModelCo <- apply(subset(SampleCorrectYWithPC1, select=-c(contig,PC1)), 2, lmColumn, y=SampleCorrectYWithPC1$PC1)
        cor_model <- apply(subset(SampleCorrectYWithPC1, select=-c(contig,PC1)), 2, function(x) cor.test(SampleCorrectYWithPC1$PC1, x)$estimate)

        estPTRs <- data.frame("estPTR"=2^abs(lmModelCo[1,]*(rangeXrm1[1]-rangeXrm1[2])), "coefficient"=lmModelCo[1,], "pValue"=lmModelCo[2,], "cor"=cor_model)
        estPTRs$sample = rownames(estPTRs)
        estPTRsEach1 <- merge(estPTRs, aggregate(correctY~sample, summeryMeanYSortFilteredSampleContigXrm1, FUN="median"), by = "sample")

        SampleCorrectYWithPC1 <- merge( dcast(subset(summeryMeanYSortFilteredSampleContigXrm2,select=c("sample", "contig", "correctY")), contig ~ sample), contigPCAPC1FilteredXrm2 )

        lmModelCo <- apply(subset(SampleCorrectYWithPC1, select=-c(contig,PC1)), 2, lmColumn, y=SampleCorrectYWithPC1$PC1)
        cor_model <- apply(subset(SampleCorrectYWithPC1, select=-c(contig,PC1)), 2, function(x) cor.test(SampleCorrectYWithPC1$PC1, x)$estimate)

        estPTRs <- data.frame("estPTR"=2^abs(lmModelCo[1,]*(rangeXrm2[1]-rangeXrm2[2])), "coefficient"=lmModelCo[1,], "pValue"=lmModelCo[2,], "cor"=cor_model)
        estPTRs$sample = rownames(estPTRs)
        estPTRsEach2 <- merge(estPTRs, aggregate(correctY~sample, summeryMeanYSortFilteredSampleContigXrm2, FUN="median"), by = "sample")

        minor_sample1 <- cor_diff(estPTRsEach1)
        minor_sample2 <- cor_diff(estPTRsEach2)
        if( (length(minor_sample1) > 0 & length(minor_sample2) > 0) | (max(estPTRsEach1$estPTR) < 1.8 & max(estPTRsEach2$estPTR) < 1.8) | (max(estPTRsEach1$estPTR)/min(estPTRsEach1$estPTR)>5 & max(estPTRsEach2$estPTR)/min(estPTRsEach2$estPTR)>5) ) {
          next
        }

        estPTRsEach12 <- merge(estPTRsEach1,estPTRsEach2,by="sample")

        if(nrow(estPTRsEach12)>0.9*nrow(estPTRsEach1) & nrow(estPTRsEach12)>0.9*nrow(estPTRsEach2)){
          cor_current <- cor(estPTRsEach12$estPTR.x,estPTRsEach12$estPTR.y)
          #print(cor_current)
          if (cor_current > max_cor) {
            max_cor <- cor_current
          }

          if( cor(estPTRsEach12$estPTR.x,estPTRsEach12$estPTR.y) > cor_cutoff ) {
            tag_permu <- 1
            estPTRsEach12$estPTR <- apply(subset(estPTRsEach12,select=c("estPTR.x","estPTR.y")),1,mean)
            estPTRsEach12$coefficient <- apply(subset(estPTRsEach12,select=c("coefficient.x","coefficient.y")),1,function(x) mean(abs(x)))
            estPTRsEach12$pValue <- apply(subset(estPTRsEach12,select=c("pValue.x","pValue.y")),1,max)
            estPTRsEach12$cor <- apply(subset(estPTRsEach12,select=c("cor.x","cor.y")),1,function(x) mean(abs(x)))
            estPTRsEach12$correctY <- apply(subset(estPTRsEach12,select=c("correctY.x","correctY.y")),1,mean)

            estPTRs2 <- subset(estPTRsEach12,select=c("sample","estPTR","coefficient","pValue","cor","correctY"))
            break
          }
        }

      }
      if(tag_permu == 1){
        break
      }

    }
    if(tag_permu == 1){
      break
    }
  }

}

```


```{r demic output}

my_pivot <- SampleCorrectYWithPC1 %>% pivot_longer(cols = c(-contig, -PC1)) 

ggplot(my_pivot) +
 aes(x = PC1, y = value) +
 geom_point(size = 1L, colour = "#0c4c8a") +
    geom_smooth(method = "lm") +
 labs(x = "Inferred relative distances to replication origin", y = "log-transformed coverages of contigs in samples", title = "DEMIC output", subtitle = "Linear models of coverage ~ contigs ordered by inferred distance to origin") +
 theme_minimal() +
 facet_wrap(vars(name), scales = "free")

```

# Linear regression fitted lines per sample

```{r}

ggplot(my_pivot) +
 aes(x = PC1, y = value, color = name) +
 # geom_point(size = 1L, colour = "#0c4c8a") +
    geom_smooth(method = "lm") +
 labs(x = "Inferred relative distances to replication origin", y = "log-transformed coverages of contigs in samples", title = "DEMIC output", subtitle = "Linear regression fitted lines per sample", color = "Day") +
 theme_minimal() +
     scale_color_brewer(palette = "Set2")
 # facet_wrap(vars(name), scales = "free")

```


And here is the estimated peak-to-trough ratio for each sample:

```{r}
estPTRs2 %>%
pander()

```

# LM of Demic output

Finally, let us test the difference in estimated peak-to-trough values:

```{r}

demicdf <- estPTRs2 %>% select(sample, estPTR) %>%
    rename(SampleID = sample)

demicands <- demicdf %>%
  left_join(sample_sheet, by = "SampleID") %>%
  rename(Condition = study_group) %>%
  select(SampleID, estPTR, Condition) %>%
  mutate(Condition = fct_relevel(Condition, "Control"))
  
ggplot(demicands, aes(x=SampleID, y=estPTR, color = Condition)) +
geom_boxplot() +
 scale_color_brewer(palette = "Set2") +
 labs(x = "Days", y = "DEMIC output (Higher values indicate more replication)", title = "DEMIC output", subtitle = "Peak-to-trough values calculated by coverage vs contigs ordered by inferred distance to contigs") +
 theme_minimal() +
    ylim(1, 3)

lm3 <- lm("estPTR ~ Condition", data = demicands)

summary(lm3)

```


